servicetalk-concurrent-api/src/test/java/io/servicetalk/concurrent/api/publisher/FromIntRangePublisherTest.java

```diff
@@ -0,0 +1,179 @@
+    }
+
+    @Test
+    public void overflowStride() {
+        int begin = Integer.MAX_VALUE - 1;
+        toSource(range(begin, Integer.MAX_VALUE, 10)).subscribe(subscriber);
+        subscriber.request(1);
+        assertThat(subscriber.takeItems(), contains(begin));
+        assertTrue(subscriber.isCompleted());
+    }
+
+    @Test
+    public void negativeToPositive() {
+        toSource(range(-2, 3)).subscribe(subscriber);
+        subscriber.request(5);
+        assertThat(subscriber.takeItems(), contains(-2, -1, 0, 1, 2));
+        assertTrue(subscriber.isCompleted());
+    }
+
+    @Test
+    public void negativeToPositiveStride() {
+        toSource(range(-1, Integer.MAX_VALUE, Integer.MAX_VALUE)).subscribe(subscriber);
+        subscriber.request(2);
+        assertThat(subscriber.takeItems(), contains(-1, Integer.MAX_VALUE - 1));
+        assertTrue(subscriber.isCompleted());
+    }
+
+    @Test
+    public void allNegative() {
+        toSource(range(-10, -5)).subscribe(subscriber);
+        subscriber.request(5);
+        assertThat(subscriber.takeItems(), contains(-10, -9, -8, -7, -6));
+        assertTrue(subscriber.isCompleted());
+    }
+
+    @Test
+    public void allNegativeStride() {
+        toSource(range(-20, -10, 2)).subscribe(subscriber);
+        subscriber.request(5);
+        assertThat(subscriber.takeItems(), contains(-20, -18, -16, -14, -12));
+        assertTrue(subscriber.isCompleted());
+    }
+
+    @Test
+    public void thrownExceptionTerminatesOnError() {
+        thrownExceptionTerminatesOnError(false);
+    }
+
+    @Test
+    public void thrownExceptionTerminatesOnErrorStride() {
+        thrownExceptionTerminatesOnError(true);
+    }
+
+    private void thrownExceptionTerminatesOnError(boolean stride) {
+        toSource((stride ? range(0, 5, 2) : range(0, 5)).whenOnNext(n -> {
+            throw DELIBERATE_EXCEPTION;
+        })).subscribe(subscriber);
+        subscriber.request(1);
+        assertThat(subscriber.takeError(), sameInstance(DELIBERATE_EXCEPTION));
+    }
+
+    @Test
+    public void reentrantDeliversInOrder() {
+        toSource(range(0, 5).whenOnNext(n -> subscriber.request(1))).subscribe(subscriber);
+        subscriber.request(1);
+        assertThat(subscriber.takeItems(), contains(0, 1, 2, 3, 4));
+        assertTrue(subscriber.isCompleted());
+    }
+
+    @Test
+    public void reentrantDeliversInOrderStride() {
+        toSource(range(5, 15, 4).whenOnNext(n -> subscriber.request(1))).subscribe(subscriber);
+        subscriber.request(1);
+        assertThat(subscriber.takeItems(), contains(5, 9, 13));
+        assertTrue(subscriber.isCompleted());
+    }
+
+    @Test
+    public void reentrantCancelStopsDelivery() {
+        reentrantCancelStopsDelivery(true);
+    }
+
+    @Test
+    public void reentrantCancelStopsDeliveryStride() {
+        reentrantCancelStopsDelivery(false);
+    }
+
+    private void reentrantCancelStopsDelivery(boolean stride) {
```

Comment:
An additional test case may be useful to verify `request-n` post `cancel` are ignored.